### ISSUE TITLE
BusinessRule with async/await

### ISSUE THREAD

**Issue by ajj7060** [MarimerLLC/cslaforum#437]
I recently discovered that a few years ago now the Dataportal now works properly with async/await, in that you can do `private async Task DataPortal_Fetch(int id)` and it works properly, especially when throwing an exception.

However, we're now doing `BusinessRule` and are using async/await there.  My concern is that the method we need to override, Execute, is void, so we end up doing `protected override async void Execute(RuleContext)`, and if an exception occurs, try/catch/finally blocks don't seem to work.. and of course we're using finally to call `context.Complete`. 

What's the recommendation when using async/await with business rules to avoid hangs or rules never completing?

**Comment by fujiiface** [comment]
I recently started reading up more on async/await programming and learned that try/catch doesn't really work with it well.  Here's some good quick videos about it.  https://channel9.msdn.com/Series/Three-Essential-Tips-for-Async.  Also, the recommendation is not to use void and async/await together.

**Comment by ajj7060** [comment]
Right, async void and try catch don't seem to work together (catch/finally doesn't execute).  But at the same time, you need to do async rules when using HttpProxy (which only supports async) or on Xamarin and similar platforms, where your app will be killed if they go to not responding.

Hence my question; the signature of Execute is void, but you *have* to do async in the rule in the above cases.  I'm tried having Execute await a method that takes a Task so try / catch works, and so far it seems to:

```
protected override async void Execute(RuleContext) => await RunRule();

private Task RunRule() {
  try {
     // some code
   } catch(Exception ex) {
     context.AddErrorResult(ex.Message);
   } finally {
     context.Complete();
   }
}
```

But is there anything else?  Are there plans to change the rule engine to have an ExecuteAsysnc, much like the DataPortal supports `private Task DataPortal_Fetch(object criteria)` ?
